src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/CertificateData.cs

```diff
@@ -59,126 +61,57 @@ internal CertificateData(byte[] rawData)
-            NotBefore = validity.ReadX509Date();
-            NotAfter = validity.ReadX509Date();
+            Subject = new X500DistinguishedName(tbsCertificate.Subject.ToArray());
 
-            if (validity.HasData)
-                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
-
-            Subject = new X500DistinguishedName(tbsCertificate.ReadNextEncodedValue());
-
-            SubjectPublicKeyInfo = tbsCertificate.ReadNextEncodedValue();
-            DerSequenceReader subjectPublicKeyInfo = new DerSequenceReader(SubjectPublicKeyInfo);
-            DerSequenceReader subjectKeyAlgorithm = subjectPublicKeyInfo.ReadSequence();
-            PublicKeyAlgorithm.AlgorithmId = subjectKeyAlgorithm.ReadOidAsString();
-            PublicKeyAlgorithm.Parameters = subjectKeyAlgorithm.HasData ? subjectKeyAlgorithm.ReadNextEncodedValue() : Array.Empty<byte>();
-
-            if (subjectKeyAlgorithm.HasData)
-                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
-
-            PublicKey = subjectPublicKeyInfo.ReadBitString();
-
-            if (subjectPublicKeyInfo.HasData)
-                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
-
-            if (Version > 0 &&
-                tbsCertificate.HasData &&
-                tbsCertificate.PeekTag() == DerSequenceReader.ContextSpecificConstructedTag1)
-            {
-                IssuerUniqueId = tbsCertificate.ReadBitString();
-            }
-            else
+            using (AsnWriter writer = AsnSerializer.Serialize(tbsCertificate.SubjectPublicKeyInfo, AsnEncodingRules.DER))
             {
-                IssuerUniqueId = null;
+                SubjectPublicKeyInfo = writer.Encode();
             }
 
-            if (Version > 0 &&
-                tbsCertificate.HasData &&
-                tbsCertificate.PeekTag() == DerSequenceReader.ContextSpecificConstructedTag2)
-            {
-                SubjectUniqueId = tbsCertificate.ReadBitString();
-            }
-            else
-            {
-                SubjectUniqueId = null;
-            }
+            PublicKeyAlgorithm.AlgorithmId = tbsCertificate.SubjectPublicKeyInfo.Algorithm.Algorithm.Value;
+            PublicKeyAlgorithm.Parameters = tbsCertificate.SubjectPublicKeyInfo.Algorithm.Parameters?.ToArray() ?? Array.Empty<byte>();
+
+            PublicKey = tbsCertificate.SubjectPublicKeyInfo.SubjectPublicKey.ToArray();
+
+            IssuerUniqueId = Version >= 1 ? tbsCertificate.IssuerUniqueId?.ToArray() : null;
+            SubjectUniqueId = Version >= 1 ? tbsCertificate.SubjectUniqueId?.ToArray() : null;
 
             Extensions = new List<X509Extension>();
 
-            if (Version > 1 &&
-                tbsCertificate.HasData &&
-                tbsCertificate.PeekTag() == DerSequenceReader.ContextSpecificConstructedTag3)
+            if (Version >= 2 && tbsCertificate.Extensions.HasValue)
```

Comment:
This if says to only copy the extensions if the version is 2 or more, and ignore them for 1 or 0.  It lost the "throw if extensions appear in version 0 or 1" part.

Since the whole structure was already parsed (as well as the original code did) this also can be deferred until something asks about it, to save memory.

Reply:
Good catch.